allennlp/modules/seq2vec_encoders/cnn_encoder.py

```diff
@@ -102,22 +106,56 @@ def forward(self, tokens: torch.Tensor, mask: torch.BoolTensor):
+        # masking, then do max pooling over each filter for the whole input sequence.
+        # Because our max pooling is simple, we just use `torch.max`.  The resultant tensor has shape
         # `(batch_size, num_conv_layers * num_filters)`, which then gets projected using the
         # projection layer, if requested.
 
+        # To ensure the cnn_encoder respects masking we add a large negative value to
+        # the activations of all filters that convolved over a masked token. We do this by
+        # first enumerating all filters for a given convolution size (torch.arange())
+        # then by comparing it to an index of the last filter that does not involve a masked
+        # token (.ge()) and finally adjusting dimensions to allow for addition and multiplying
+        # by a large negative value (.unsqueeze())
         filter_outputs = []
+        batch_size = tokens.shape[0]
+        # shape: (batch_size, 1)
+        last_unmasked_tokens = mask.sum(dim=1).unsqueeze(dim=-1)
         for i in range(len(self._convolution_layers)):
             convolution_layer = getattr(self, "conv_layer_{}".format(i))
-            filter_outputs.append(self._activation(convolution_layer(tokens)).max(dim=2)[0])
+            pool_length = tokens.shape[2] - convolution_layer.kernel_size[0] + 1
+
+            # Forward pass of the convolutions.
+            # shape: (batch_size, num_filters, pool_length)
+            activations = self._activation(convolution_layer(tokens))
+
+            # Create activation mask.
+            # shape: (batch_size, pool_length)
+            indices = torch.arange(pool_length).unsqueeze(0).expand(batch_size, pool_length)
+            # shape: (batch_size, pool_length)
+            activations_mask = indices.ge(
+                last_unmasked_tokens - convolution_layer.kernel_size[0] + 1
+            )
+            # shape: (batch_size, num_filters, pool_length)
+            activations_mask = activations_mask.unsqueeze(1).expand_as(activations)
+
+            # Replace masked out values with smallest possible value of the dtype so
+            # that max pooling will ignore these activations.
+            # shape: (batch_size, pool_length)
+            activations = activations + (activations_mask * min_value_of_dtype(activations.dtype))
+
+            # Pick out the max filters
+            filter_outputs.append(activations.max(dim=2)[0])
 
         # Now we have a list of `num_conv_layers` tensors of shape `(batch_size, num_filters)`.
         # Concatenating them gives us a tensor of shape `(batch_size, num_filters * num_conv_layers)`.
         maxpool_output = (
             torch.cat(filter_outputs, dim=1) if len(filter_outputs) > 1 else filter_outputs[0]
         )
 
+        # Replace the maxpool activations that picked up the masks with 0s
+        maxpool_output[maxpool_output == min_value_of_dtype(maxpool_output.dtype)] = 0.0
```

Comment:
One more question: why set the masked-out activations to 0 here, as opposed to, say, `min_value_of_dtype`?

Reply:
These already would be the min value of dtype, if we pass this in then a bunch of model tests will fail since those values get propagated to model components downstream resulting in nan losses etc. Initially I thought we should actually throw errors if this happens but i have not considered character level models, for which this will happen pretty much always.